### PR TITLE
fix: LEAP-1490: Fix AudioUltra working with relations and comments

### PR DESCRIPTION
**Pull request by Gondragos**
This PR fixes the way AudioUltra works with Relations and linking comments (it didn't work and that was a regression), and some small bugs related to AudioUltra itself.
 
- [getting AudioRegion bbox](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-90c91b60046b0b57fdfc7d81075411f883a078c3df01aa68e3e048693c69a694R117-R132) based on in-model values 
 - reusing [propertyWatcher](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-19130085f2d864c56833f4e42111d0fee1c0a61f7cd0d4bd3a8c6e4cf42a9eabR26) to observe changes in onscreen region position
- adding [active](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-814acb205a470a72381b3ae5a9b57845299e5e6486e210281754cd67763c2971R51-R52) state for the region to separate it from highlighted one to prevent them canceling each other
- moving [in-canvas region y-coordinates](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-814acb205a470a72381b3ae5a9b57845299e5e6486e210281754cd67763c2971L329-L331) to [getters](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-814acb205a470a72381b3ae5a9b57845299e5e6486e210281754cd67763c2971R173-R186) and fixing its height (taking into account timeline size)
- fixing inViewport [conditions](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-814acb205a470a72381b3ae5a9b57845299e5e6486e210281754cd67763c2971R223). viewport might be based on the canvas size only as it's the real element's frame.
- adding inner event on Waveform to indicate current visible frame state on each redraw
- [using that event](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-a8735c7089bda3857898ddaf991432c18c997c736b16d0f89d7c47e6683c6510R33-R55), storing the [frame state in the model](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-89e98192c0f9010208c26dda9f397621af1b42408bf533864896a921080b1d6eR61-R63) [to catch the moment](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-a363ff118e1d49e4001ddfdaa8e659e89cfcc94fb56a405a50a8aaeb510d2e18R24) when we [need recalculating region  bbox](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-a363ff118e1d49e4001ddfdaa8e659e89cfcc94fb56a405a50a8aaeb510d2e18R24)
- adding [on-region listeners](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-a363ff118e1d49e4001ddfdaa8e659e89cfcc94fb56a405a50a8aaeb510d2e18R173-R176) to apply highlightability
- integrating [linking modes logic](https://github.com/HumanSignal/label-studio/pull/6494/files#diff-89e98192c0f9010208c26dda9f397621af1b42408bf533864896a921080b1d6eR104-R110) on selecting audioultra region
- [Fix sticking in highlighted status](https://github.com/HumanSignal/label-studio/pull/6494/commits/d99a7c8ad5296c3d45dd78a412f8d66e0606f186)

### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [x] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
_(check all that apply)_
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend

#### What feature flags were used to cover this change?
N/A

### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [x] Not sure (briefly explain the situation below)

It shouldn't be like that if I don't break something.

### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [x] integration
- [ ] unit

### Which logical domain(s) does this change affect?
`Audio`, `AudioUltra`, `Relations`, `LinkingModes`, `Comments`

### CODE REVIEW

File: web/libs/editor/src/tags/object/AudioUltra/model.js [hunk 1]
```diff
@@ -131,6 +132,9 @@ export const AudioModel = types.compose(
     })
     .volatile(() => ({
       errors: [],
+      stageRef: createRef(),
+      _ws: null,
+      _wf_frame: null,
```
Comment (bmartel) [inline web/libs/editor/src/tags/object/AudioUltra/model.js]:
```suggestion
      _wfFrame: null,
```

File: web/libs/editor/src/tags/object/AudioUltra/model.js [hunk 2]
```diff
@@ -534,6 +538,9 @@ export const AudioModel = types.compose(
             console.warn("Already destroyed");
           }
         },
+        setWFFrame(frame) {
+          self._wf_frame = frame;
```
Comment (bmartel) [inline web/libs/editor/src/tags/object/AudioUltra/model.js]:
```suggestion
          self._wfFrame = frame;
```

File: web/libs/editor/src/regions/AudioRegion/AudioUltraRegionModel.js [hunk 3]
```diff
@@ -16,8 +17,30 @@ export const AudioUltraRegionModel = types
   })
   .volatile(() => ({
     hideable: true,
+    _ws_region: null,
   }))
   .views((self) => ({
+    get bboxTriggers() {
+      return [self.start, self.end, self._ws_region, self.object?._ws, self.object?._wf_frame];
```
Comment (bmartel) [inline web/libs/editor/src/regions/AudioRegion/AudioUltraRegionModel.js]:
```suggestion
      return [self.start, self.end, self._wsRegion, self.object?._ws, self.object?._wfFrame];
```

File: web/libs/editor/src/tags/object/AudioUltra/model.js (hunk 1 above)
Comment (bmartel) [reply web/libs/editor/src/tags/object/AudioUltra/model.js]: I know there was historically a `_ws_region` which likely influenced this naming, but that was a mistake it should have been camelCase not snake_case. I don't know how ingrained `_ws_region` is already, but if at all possible could we rename both of these to camelCase like `_wfFrame` and `_wsRegion`.

File: web/libs/editor/src/regions/AudioRegion/AudioUltraRegionModel.js [hunk 4]
```diff
@@ -143,5 +166,14 @@ export const AudioUltraRegionModel = types
           self.updatePosition();
         }
       },
+
+      setWSRegion(wsRegion) {
+        self._ws_region = wsRegion;
```
Comment (bmartel) [inline web/libs/editor/src/regions/AudioRegion/AudioUltraRegionModel.js:171]:
```suggestion
        self._wsRegion = wsRegion;
```

File: web/libs/editor/src/regions/AudioRegion/AudioUltraRegionModel.js [hunk 5]
```diff
@@ -16,8 +17,30 @@ export const AudioUltraRegionModel = types
...
+      return [self.start, self.end, self._ws_region, self.object?._ws, self.object?._wf_frame];
+    },
+    get bboxCoordsCanvas() {
+      if (!self.bboxTriggers) {
+        return null;
+      }
+
+      const { _ws_region } = self;
+      if (!_ws_region) return null;
+      if (!_ws_region.inViewport) return null;
```
Comment (bmartel) [inline web/libs/editor/src/regions/AudioRegion/AudioUltraRegionModel.js:33]:
```suggestion
      const { _wsRregion } = self;
      if (!_wsRegion) return null;
      if (!_wsRegion.inViewport) return null;
```

**Review by bmartel — APPROVED** [review]
Looks great 😎, just a suggestion on the snake_case -> camelCase, but logically speaking that won't affect this getting tested. Accidentally approved prematurely, taking this back to when the PR is considered for review officially.

**Comment by Gondragos** [comment]
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/11366932003)
> Successfully merged: Already up to date.

**Comment by robot-ci-heartex** [comment]
Follow Merge downstream workflow has been failed.
> [Workflow run](https://github.com/HumanSignal/label-studio-enterprise/actions/runs/11366921419)

**Comment by Gondragos** [comment]
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/11384132396)
> Successfully merged:  21 files changed, 386 insertions(+), 224 deletions(-)

File: web/libs/editor/src/regions/AudioRegion/AudioUltraRegionModel.js [hunk 6]
```diff
@@ -16,8 +17,30 @@ export const AudioUltraRegionModel = types
   })
   .volatile(() => ({
     hideable: true,
+    _ws_region: null,
```
Comment (Gondragos) [inline web/libs/editor/src/regions/AudioRegion/AudioUltraRegionModel.js:20]:
```suggestion
    _wsRegion: null,
```

File: web/libs/editor/src/regions/AudioRegion/AudioUltraRegionModel.js [hunk 7]
```diff
@@ -16,8 +17,30 @@ export const AudioUltraRegionModel = types
...
+    get bboxCoordsCanvas() {
+      if (!self.bboxTriggers) {
+        return null;
+      }
+
+      const { _ws_region } = self;
+      if (!_ws_region) return null;
+      if (!_ws_region.inViewport) return null;
+
+      const { xStart, xEnd, yStart, yEnd, visualizer } = _ws_region;
```
Comment (Gondragos) [inline web/libs/editor/src/regions/AudioRegion/AudioUltraRegionModel.js:35]:
```suggestion
      const { xStart, xEnd, yStart, yEnd, visualizer } = _wsRegion;
```

File: web/libs/editor/src/tags/object/AudioUltra/model.js (hunk 1 above)
Comment (Gondragos) [reply web/libs/editor/src/tags/object/AudioUltra/model.js]: @bmartel I've returned to the `_ws_region` as there are a lot of places which it affects. In general, the worst part is that we have three versions of the `Audio` tag and if we change it here we will need to change it everywhere. I prefer not to touch it until we remove the related feature flags and legacy versions.

File: web/libs/editor/src/lib/AudioUltra/Regions/Segment.ts [hunk 8]
```diff
@@ -168,6 +170,21 @@ export class Segment extends Events<SegmentEvents> {
...
+    const top = timelineLayer?.isVisible && timelineTop ? timelineHeight : 0;
+
+    return top;
+  }
+
+  get yEnd() {
+    const { height } = this.visualizer;
+    const { timelineHeight } = this;
+    return this.yStart + (height - timelineHeight);
+  }
```
Comment (bmartel) [inline web/libs/editor/src/lib/AudioUltra/Regions/Segment.ts:186]: Thank you, this is a nice refactor that makes the readability of the code much better.

File: web/libs/editor/src/lib/AudioUltra/react/index.ts [hunk 9]
```diff
@@ -29,6 +30,30 @@ export const useWaveform = (
...
+          frameState.height !== prevFrameState.height ||
+          frameState.zoom !== prevFrameState.zoom ||
+          frameState.scroll !== prevFrameState.scroll
+        ) {
+          onFrameChangedRef.current?.(frameState);
+          prevFrameState = frameState;
+        }
+      });
+    };
+    return updateAfterRegionDraw;
```
Comment (bmartel) [inline web/libs/editor/src/lib/AudioUltra/react/index.ts]:
```suggestion
    return (frameState: WaveformFrameState) => {
      cancelAnimationFrame(requestId);
      requestId = requestAnimationFrame(() => {
        if (
          !prevFrameState ||
          frameState.width !== prevFrameState.width ||
          frameState.height !== prevFrameState.height ||
          frameState.zoom !== prevFrameState.zoom ||
          frameState.scroll !== prevFrameState.scroll
        ) {
          onFrameChangedRef.current?.(frameState);
          prevFrameState = frameState;
        }
      });
    };
```
Just so it doesn't shadow the main defined const name, which can look on first glance like it is being redefined.

File: web/libs/editor/tests/integration/e2e/audio/audio_regions.cy.ts [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    // deactivating
+    AudioView.seekCurrentTimebox(0);
+    const inactiveRegionColor = AudioView.getPixelColorRelative(0.36, 0.9);
+
+    inactiveRegionColor.then((color) => {
+      baseRegionColor.should("deep.equal", color);
+    });
+  });
+
+  it("Should have indication of highligted state", () => {
```
Comment (bmartel) [inline web/libs/editor/tests/integration/e2e/audio/audio_regions.cy.ts]:
```suggestion
  it("Should have indication of highlighted state", () => {
```

**Review by bmartel — APPROVED** [review]
Amazing work 🔥

**Comment by Gondragos** [comment]
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/11441710066)
> Successfully merged:  delete mode 100644 web/libs/datamanager/src/components/Label/Toolbar/Toolbar.scss

**Comment by robot-ci-heartex** [comment]
Follow Merge downstream workflow has been failed.
> [Workflow run](https://github.com/HumanSignal/label-studio-enterprise/actions/runs/11442357488)

**Comment by robot-ci-heartex** [comment]
Follow Merge downstream workflow has been failed.
> [Workflow run](https://github.com/HumanSignal/label-studio-enterprise/actions/runs/11442357488)

**Comment by robot-ci-heartex** [comment]
Follow Merge downstream workflow has been failed.
> [Workflow run](https://github.com/HumanSignal/label-studio-enterprise/actions/runs/11442357488)

**Pull request merged**